### ISSUE TITLE
Git support

### ISSUE THREAD

**Issue by jpallen** [overleaf/overleaf#10]
The versioning api already uses git behind the scenes, so we can extend it to allow git push and pull.

Issues/stepping stones:
- The versioning API is old code and does not match the standards of the rest of our services
- Handling conflicts - i.e. A push when there are uncommitted changes either in web or the git repo in the versioning API.
- Incoming pushes must send their changes to the third party data store.
- Incoming pushes/pulls must be sent via an ssh proxy which will perform auth and ensure only the authorised repositories can be accessed.

**Comment by jpallen** [comment]
On www.sharelatex.com you can now push and pull your project to a GitHub repo.

**Comment by ffurrer** [comment]
Is GitHub integration also available when you host sharelatex yourself?

**Comment by flop25** [comment]
no, they have made a clear sepration between opensource features/code and
not opensource one

2015-01-22 16:50 GMT+01:00 Fadri Furrer <notifications@github.com>:

> Is GitHub integration also available when you host sharelatex yourself?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sharelatex/sharelatex/issues/10#issuecomment-71041511>
> .
>



-- 
__

**Comment by okennedy** [comment]
+1 for being able to do a git push/pull on a sharelatex repo.

**Comment by liob** [comment]
+1 for being able to do a git push/pull on a sharelatex repo.

Overleaf now provides push/pull support without authorization for unlisted projects.

**Comment by ffurrer** [comment]
but they don't support tracking the current status from the website on a git repo on e.g. github, as far as I understood it (didin't test anything), but only support pulling and pushing to the website from a local git repo.

**Comment by moschlar** [comment]
I would guess that people who run their own Sharelatex instance (e.g. for privacy reasons) most likely aren't too interested in GitHub(-only) integration either.
Being able to use Sharelatex as a git remote from my local machine would be already very pleasant for me and probably not too complicated to implement. If Sharelatex were to push (and pull) on its own to other remotes, it would definitely get more complicated.

So, :+1: from me for "plain" git support.

**Comment by dnano91** [comment]
git support would really be awesome

**Comment by LizardM4** [comment]
:+1: for git support. I'd love to pair sharelatex with gitlab, but raw support would as well be awesome. Looking forward to see this big feature.

**Comment by eikelangbehn** [comment]
Git support would be really great!

**Comment by rubdos** [comment]
I think it shouldn't be too difficult to integrate with an existing Git(Lab) server (or any other git-over-ssh supporting server), as the code already exists for the commercial/non-free sharelatex.

Anyhow, do NOT do it like overleaf does. That's aweful.

That said, I thought this already was a feature of ShareLaTeX. Please inform

> *    The versioning API is old code and does not match the standards of the rest of our services

Is this still the case?

> *    Handling conflicts - i.e. A push when there are uncommitted changes either in web or the git repo in the versioning API.

Overleaf solves this by commiting uncommit changes (with commit message "Update on OverLeaf") in web when someone accesses the git repo. I repeat: do this in a sane way, please.

> *    Incoming pushes must send their changes to the third party data store.

That could be GitLab/any ssh-git server

> *    Incoming pushes/pulls must be sent via an ssh proxy which will perform auth and ensure only the authorised repositories can be accessed.

That could be GitLab.

I would follow the Unix philosophy here. Do not implement a whole git backend by yourself; rather integrate with wat already exists.

I would be glad helping, if someone could guide me through the code base.

**Comment by arkhan** [comment]
+1

**Comment by jlazerus** [comment]
+1

On Fri, Oct 30, 2015 at 8:37 AM, Edison Ibañez <notifications@github.com>
wrote:

> +1
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sharelatex/sharelatex/issues/10#issuecomment-152542800>
> .
>



-- 
Jeff Lazerus
http://www.jefflazerus.com/

**Comment by PunKeel** [comment]
+1 :)

**Comment by rubdos** [comment]
Come on guys, some feedback please. I'd love to help in the integration, but the last comment from ShareLaTeX has been a year ago.

**Comment by henryoswald** [comment]
We are not working or planning this at the moment.

On Saturday, 19 December 2015, Ruben De Smet <notifications@github.com>
wrote:

> Come on guys, some feedback please. I'd love to help in the integration,
> but the last comment from ShareLaTeX has been a year ago.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sharelatex/sharelatex/issues/10#issuecomment-166021860>
> .
>


-- 
Henry Oswald
shareLaTeX.com co-founder

**Comment by rubdos** [comment]
Is it possible to start planning this? It's kind of a killer feature which makes people stay at Overleaf at the moment. I'd love to help.

**Comment by har-wradim** [comment]
+1, @rubdos formulated the idea very neatly

**Comment by henryoswald** [comment]
yeah @rubdos comments were very helpful, sorry for not helping clarify some of the questions.

- The versioning API is now dead so that can be ignored.
- The biggest issue is how do we manage a git backend? We don't want to have a different runtime (gitlab) in the stack as a dependency really.
- Do we have to have a filesystem base git or is there some other implementation we can use to make life simpler?
- If we do need a filesystem git how can we make it redundant and scalable? Like all our datastores it needs to be able to take the total loss of a node.
- The authentication part I'm not too worried about, its not simple but I think there are some obvious answers for it.

It could be a case of we don't have to manage the git repos traditionally. We store them in s3/filestore and pull them down on a push.

**Comment by rubdos** [comment]
>  The versioning API is now dead so that can be ignored.

Roger that

>    The biggest issue is how do we manage a git backend? We don't want to have a different runtime (gitlab) in the stack as a dependency really.
>    Do we have to have a filesystem base git or is there some other implementation we can use to make life simpler?
>    If we do need a filesystem git how can we make it redundant and scalable? Like all our datastores it needs to be able to take the total loss of a node.
>    The authentication part I'm not too worried about, its not simple but I think there are some obvious answers for it.

> It could be a case of we don't have to manage the git repos traditionally. We store them in s3/filestore and pull them down on a push.

Can I suggest something that's pretty much completely different?

I would love to see ShareLaTeX as 'just' the editor. So when git is enabled for a project, ShareLaTeX makes sure that only compilable stuff gets into a backend git (which, then, can be anything from a stupid gitweb to GitHub to GitLab self hosted). It could even be on a per user basis (forks?).

Workflow would then be

* fork of main git repo
* make changes
* commit and push

If last fails, you'll have to solve the merge; but that can be done in the web frontend.

I think that would hook up pretty well in the git workflow many people have when using LaTeX. The UI can make it very easy to do merges well.

I'd give ShareLaTeX its own SSH key to do push and pull, which you have to set once per repository. I don't think ShareLaTeX needs to have it's own git backend though. That would make things overly complicated imo.

Please note that I'm not fully awake at the moment, so if I'm not clear, please tell me. Thank you for looking into this.

**Comment by henryoswald** [comment]
Hi,

I think we are talking at cross purposes a little bit here. Your comment was more about the user workflow but there is a fundamental issue we need to solve of when a user doe a git push ssh://git.sharelatex.com/etc where does this go? ShareLaTeX is a http service which persists data in mongodb, redis and s3 running on a collection of stateless machines. We don't have a place where a load of git repos can live happily forever at the moment which is why I mentioned compressing them and pushing them up to s3.

**Comment by LizardM4** [comment]
I'm getting confused. I think here there are two ways people intend "git support":
  1. Push from your local repo to a ShareLaTeX-hosted (whatever that means) repo, and have ShareLaTeX see the changes.
  2. Have the repository hosted on some other service, and have ShareLaTeX push to/pull from it.

Correct me if I'm wrong, but I think that @rubdos means (2) and instead @henryoswald is thinking about (1). I also had (2) in mind: wouldn't then a new pulled commit be stored just as a user edit? (here I'm deliberately forgetting merge conflicts, rewritten history & so on...)

**Comment by henryoswald** [comment]
ah right so are you saying there will be no 'git push ssh://git.sharelatex.com' but 'git push ssh://git.bitbucket/gitlab/generic.com'  ?

**Comment by LizardM4** [comment]
Yes, but I am not acquainted with the code, so I don't know if it's doable. The idea is: let's leave git to git services, and do just document editing. It doesn't even need to handle conflict resolution. If the user leaves the branch in a way that ShareLaTeX can't push conflict-free, let's just push to a different branch and let the user deal with it.

But yeah, this is my personal view!

**Comment by henryoswald** [comment]
Cool, your concerns are the same as we have. 

I think if we push forward with this then we would want todo option 1 though as this would likely satisfy most demand.

**Comment by rubdos** [comment]
Option 1 is probably more difficult though, as you'll have to get ShareLaTeX to run a git repo together with what it's already doing. I'm not acquainted with code code either, so I'm only giving suggestions.

**Comment by henryoswald** [comment]
It is unfortunately which is why we have not tackled it yet and I think option 2 won't make that many people happy.

**Comment by rubdos** [comment]
If you integrate well with an existing git hosting solution (pure pushing and pulling) and make it easy to integrate both services, I don't think it'll be to difficult to sort of 'emulate' option 1.

I think that, when option 2 is well implemented, it's not too easy to use it like option 1 :)

**Comment by LizardM4** [comment]
Ok now I see your point better: a user that wants to work offline and online would rather not have to set up both the repo and ShareLaTeX, given that ShareLaTeX already provides storage. But on the other side, how do you implement option 1 without having ShareLaTeX to join the run for becoming a git provider?

Option 1 is hard. What about option 2, would it be difficult to set up? Just trying to look at things in perspective.

**Comment by okennedy** [comment]
I +1ed this ages ago for several reasons.

1) I work with a lot of people who don't normally use LaTeX.  A visual editor like ShareLaTeX has been incredibly helpful in making collaborations efficient.  I also work with a lot of people who have their own highly tuned personal workflows and find the ShareLaTeX UI a hindrance.  Even having git **client** support (option 2) in ShareLaTeX would be a great way to get these two types of people in sync.

2) Diagrams, graphs, figures, etc... Even if I write the diagram generator in LaTeX itself, as suggested in the ShareLaTeX docs (a process that's more convoluted than it needs to be), going through the web interface to upload data files is an annoying workflow breaker.  Git support would allow me to use my desktop to render and tweak graphs individually, and allow me to export them to ShareLaTeX when they're ready.  Repeatedly dragging and dropping into the web interface is exactly the type of file synchronization cludge that GIT and other forms of SCM were meant to avoid.

3) Archival.  I want a copy of the files under my own control once a manuscript is finished.  Downloading ZIP files is a cludge. 

In short, while Option 1 would be ideal, Option 2 would *still* be incredibly helpful.

**Comment by rubdos** [comment]
Yes, I don't see another reason to have git support. People who don't want to use the ShareLaTeX frontend themselves, but want to collaborate with people that fear tools like `vim`, `make` and commandline. Or even people that are afraid of installing a LaTeX distribution, which incorporates ~5 GB of data. Or people with Chromebooks.

While I think option 1 is still vital, I still think it will follow automagically from well implementing option 2. After all, when option 2 is implemented correctly, using a git client only differs in the URL used for fetch and push. Then ShareLaTeX should 'just' use that repo as storage, instead of whatever it does now.

While in principle that sounds easy, I can imagine that it's pretty difficult overall. In any case, I see a lot of potential when it's implemented right. Especially for students, of which some are powerusers and some never touched something else then MS Word.

To put things in perspective: my first LaTeX workflow at the uni was on a shared ownCloud instance, because people don't want to learn git. It was horrible.

**Comment by henryoswald** [comment]
@LizardM4 we have to work out a way which we can manage git projects at scale, there are lots of people who worry about this problem at github for us but for ShareLaTeX we are not experts. For example. How do we scale it so we can have 100,000 git repos inside ShareLaTeX?

Option 2 would not be super hard but its not easy, still a few weeks work imo.

**Comment by rubdos** [comment]
By the way, doesn't the sharelatex.com have Git functionalities? If so, how does that work?

**Comment by LizardM4** [comment]
Synchronisation to GitHub, not to a generic git repo, and private repos are not free here. And besides, I think the whole git idea was to be somehow decentralised. I run my personal version of gitlab, and I've seen it set up in several production environments; and you can still use a git repo just via ssh. I originally +1'd this enhancement request when I found out that I couldn't sync it to something else than GitHub.

**Comment by victorsndvg** [comment]
:+1:  For me, synchronization with a custom git server like GitLab is a must

**Comment by dfelder** [comment]
Perhaps it is possible to keep the repos in one of the databases. This article talks about using libgit2 backends (eg. with Redis) for the object and ref datastores. 
http://blog.deveo.com/your-git-repository-in-a-database-pluggable-backends-in-libgit2/

**Comment by rubdos** [comment]
Why reinvent the wheel by hosting your own git? I'd say, just use something that exists. You can even test with gitlab.com, as they offer free private repositories.

**Comment by chfo** [comment]
Why so complicated.

I was thinking to hack the gitsupport into sharelatex by myself, but there is only the half src published. The complete github src is missing (didn't find it). 

But i was thinking about this simple Workflow:
-> generate a SSH Keypair if needed for ssh git(or import keys does'nt matter but for security reasons the keys should only be used with sharelatex)
-> Start new Project via git clone <ssh/url>
-> Menupoints: git: pull, push etc. 

The handling should be nearly the same like github. It doesn't matter at a specific pint if you get the files from github or anyother git instance

**Comment by henryoswald** [comment]
ShareLaTeX's github.com integration is done entirely over http, it doesn't talk the git protocol at all.

**Comment by trosendal** [comment]
Nice discussion on this subject. I think the key to git integration is to not hide the details of git from the webeditor users. The only platform that I have seem that has nailed the webediting of a file in a git repository is github. Several examples of getting it wrong have tried to hide the details:

1. Overleaf: The site allows you to pull and push to and from the repo. The web editor makes commits regularly and upon a push from another repo. Therefore when you push from a local at the time that someone else is web editing you are always behind the head of the remote. Mistake: try to hide commiting and merging from the web user.

2. Gitlab: Edit a file and commit it in the web browser. If a push is made with changes to the same file during web editing the web editors changes overwrite the pushed changes in the previous commit. Mistake: Try to hide merging from the web user.

In my opinion the right way to do this is make every step of the web editing process (almost) transparently follow the git workflow:

* Want to edit? Click edit(branch). In the background a new branch is creating on the server. You edit the file. 
* Want to save? Click commit. A commit is made to the new branch.
* Want to share your work with others? Click merge... 
    * If there are no conflicts, git just merges the changes into the master.
    * If you are behind with conflicts, the webtool asks you if you want to resolve conflicts. It shows you git's output and perhaps highlights in color the places that need your attention. Then you click commit again.
    * You could also answer no to the resolve conflict question in which case the commits become a pull request to the owner of the project.

**Comment by npalix** [comment]
That's indeed the behavior I'm expecting too.
Maybe use the username as the edit branch name. So multiple edits can occur
concurrently, and the possible pull request can refer to an user.

**Comment by rubdos** [comment]
I agree with @trosendal. On top of that, if

> Gitlab: Edit a file and commit it in the web browser. If a push is made with changes to the same file during web editing the web editors changes overwrite the pushed changes in the previous commit. Mistake: Try to hide merging from the web user.

this is true, I'll have to open a ticket at the gitlab issue tracker.

**Comment by trosendal** [comment]
Have a look at this issue in the gitlab issue tracker:
https://gitlab.com/gitlab-org/gitlab-ce/issues/503#

**Comment by sebhahn** [comment]
What is the current status of the git(lab) support?

**Comment by sebhahn** [comment]
Anyone?

**Comment by henryoswald** [comment]
we are not working on this at the moment.

**Comment by sebhahn** [comment]
Thanks for your feedback @henryoswald! 

However, is there any interface (AP) to access the git repositories of projects? Or is it possible to access a repo via command line? (I'm talking here about a local installation of sharelatex)

**Comment by michalwidera** [comment]
Anybody knows is there any fork with git support on github ?

**Comment by zeigerpuppy** [comment]
I'm wondering if there's any progress on git(lab) integration.  option 2 above definitely sounds sufficient.
the main problem this would solve for my users is when they go offline.  So if a user is working on a document in a remote location they can pull to a local editor and then push to sync up their changed when they return.
I agree with previous comments that this feature would also allow much greater adoption by power users who want to collaborate but retain elements of their local workflow.
Any updates on a timeline for option 2.  I'd be happy to help with testing or workibg on API if we have a few pointers.

**Comment by arjunsuresh** [comment]
Is it possible to get git API working on the opensource version of sharelatex?

**Comment by rfabbri** [comment]
Git should be directly embedded in Sharelatex. Clone, push/pull directly. NO github. Why are you doing this half-heartedly?

**Comment by bkietz** [comment]
@rfabbri Whenever possible I try to convince less tech savvy colleagues/acquaintances that non-WYSIWYG authoring is a worthwhile investment of time and effort. ShareLaTeX is an extremely accessible interface but it has still been too much overhead for many to whom I've shown it. Adding the full complexity of source control would compound the problem for people who are already intimidated. I agree that full Git integration would be a worthwhile hidden feature, but also: have mercy on the non-devs. They're up here too

**Comment by stefnotch** [comment]
I wrote a little ShareLaTeX downloader with caching support. Then, you can use a local Git repository to keep your files backed up.

<https://github.com/stefnotch/sharelatex-project-sync>

**Comment by jdleesmiller** [comment]
Since this ticket has been around for a long time, and we do now have git support (https://www.overleaf.com/blog/195-new-collaborate-online-and-offline-with-overleaf-and-git-beta), I'll close it out.

We will also look into whether we can support git in a future release of the open source edition of Overleaf (which is the successor to the open source edition of ShareLaTeX).

**Comment by zeigerpuppy** [comment]
@jdleesmiller, I don't really understand how this makes sense to close as an issue when the feature is not available in this version (this repo). It's fairly important for those basing off the open source repo for the issues to be flagged if they are still present.

**Comment by jdleesmiller** [comment]
@zeigerpuppy Thanks for the feedback. I would fine with someone opening up a new issue about git support in the open source community edition. I just think we should recognise that progress has been made here and put this 5+ year old issue to bed.

I can also add that the git bridge is itself already open source: https://github.com/overleaf/writelatex-git-bridge . However, it relies on a history storage API from Overleaf that isn't (yet (I think)) open source or able to run outside of cloud environments. On overleaf.com we have tied the two storage systems together to make git support work, but we have some work ahead of us before it will be possible to bring it on premise.
